packages/docusaurus-tsconfig/tsconfig.json

```diff
@@ -0,0 +1,23 @@
+{
+  "$schema": "https://json.schemastore.org/tsconfig",
+  "display": "Docusaurus",
+  "docs": "https://docusaurus.io/docs/typescript-support",
+  "compilerOptions": {
+    "allowJs": true,
+    "esModuleInterop": true,
+    "jsx": "preserve",
+    "lib": ["DOM"],
+    "moduleResolution": "Node16",
```

Comment:
We are considering making it an error to specify `moduleResolution: "node16"` without `"module": "node16"` at https://github.com/microsoft/TypeScript/pull/54567. `"module": "node16"` implies `"moduleResolution": "node16"`, but unfortunately not the reverse. Using one without the other matching has some very strange, unpredictable, and dangerous effects.

Many of the new errors that we detected would occur from this change are in tsconfigs that inherit from the old `@tsconfig/docusaurus` package. Would you be able to add `"module": "Node16"` to this template?

Reply:
The `moduleResolution` should be `bundler`, because this config is shared for our users, who exclusively write client code. We use `"moduleResolution": "node16"` here so that package.json `exports` can be resolved, but if we use `"module": "node16"`, then because users never specify `"type": "module"` in their package.json (this actually breaks webpack, for some reason yet to be discovered), they will not be able to import ESM dependencies.

Reply:
Sure! I don't think anyone has played with the tsconfig since the last time I added `node16`. I'm a bit caught up with MDN work recently so I may not be able to come back to it, but I (or @slorber) will surely play with it when we have time.

Reply:
How do you version @docusaurus/tsconfig? Would this have to wait until a major version bump of other docusaurus packages?

Reply:
@docusaurus/tsconfig is already created for the next major, so before its release, we are free to do anything with it :) Do you anticipate the switch to `moduleResolution: bundler` to be a breaking change?

Reply:
Technically it could introduce new TS errors. With `node16` and no `"type": "module"`, resolution of imports of external libraries uses the `types`, `node`, and `require` conditions when lookup up conditional exports, because the assumption is that these imports will be emitted as `require` calls (this is part of why mixing `--moduleResolution node16` with `--module esnext` breaks some core assumptions in the compiler). With `bundler`, the conditions are `types` and `import`. This better matches what Webpack is actually going to do during its own resolution, but we do occasionally see external libraries that have messed up the types for their `import` entrypoint.

So, it’s really like a bugfix, but it could trigger an external library’s typing bug in a way that the current config doesn’t.

Reply:
I see. It should be okay for us. Our versioning scheme typically allows "more type errors that are undeniably user bugs" across minor versions.

Reply:
Hi @andrewbranch 

Our CI started failing after 5.2 GA release, so it's a good reminder for me to handle this 😅 

This will be fixed in https://github.com/facebook/docusaurus/pull/9258

Thanks